File: implementations/php/check.php
<?php

declare(strict_types=1);

require __DIR__ . '/cid.php';

$entries = scandir(CIDS_DIR);
if ($entries === false) {
    fwrite(STDERR, "Unable to read cids directory.\n");
    exit(1);
}

sort($entries);

$mismatches = [];
$download_failures = [];
$count = 0;
$base_url = 'https://256t.org';

foreach ($entries as $entry) {
    if ($entry === '.' || $entry === '..') {
        continue;
    }

    $path = CIDS_DIR . DIRECTORY_SEPARATOR . $entry;
    if (is_dir($path)) {
        continue;
    }

    $local_content = file_get_contents($path);
    if ($local_content === false) {
        $mismatches[] = "$entry could not be read";
        continue;
    }

    $count++;
    $expected = compute_cid($local_content);
    if ($expected !== $entry) {
        $mismatches[] = sprintf('%s should be %s', $entry, $expected);
    }
    
    // Check downloaded content
    try {
        $result = download_cid($base_url, $entry);
        if (!$result['is_valid']) {
            $download_failures[] = [$entry, $result['computed']];
        } elseif ($result['content'] !== $local_content) {
            $download_failures[] = [$entry, 'content mismatch with local file'];
        }
    } catch (Exception $e) {
        $download_failures[] = [$entry, $e->getMessage()];
    }
}

$has_errors = false;

if ($mismatches) {
    echo "Found CID mismatches:\n";
    foreach ($mismatches as $message) {
        echo "- {$message}\n";
    }
    $has_errors = true;
}

if ($download_failures) {
    fwrite(STDERR, "Found download validation failures:\n");
    foreach ($download_failures as [$cid, $error]) {
        fwrite(STDERR, "- {$cid}: {$error}\n");
    }
    $has_errors = true;
}

if ($has_errors) {
    exit(1);
}

echo "All {$count} CID files match their contents.\n";
echo "All {$count} downloaded CIDs are valid.\n";

File: implementations/dart/check.dart
import 'dart:io';
import 'dart:typed_data';

import 'cid.dart';

Future<void> main() async {
  final entries = cidsDir
      .listSync()
      .whereType<File>()
      .toList()
    ..sort((a, b) => a.path.compareTo(b.path));

  final mismatches = <String>[];
  final downloadFailures = <String>[];
  const baseUrl = 'https://256t.org';

  for (final file in entries) {
    final cid = file.uri.pathSegments.last;
    final localContent = file.readAsBytesSync();
    final expected = computeCid(localContent);
    
    // Check local CID file
    if (expected != cid) {
      mismatches.add('$cid should be $expected');
    }
    
    // Check downloaded content
    try {
      final result = await downloadCid(baseUrl, cid);
      if (!result.isValid) {
        downloadFailures.add('$cid: ${result.computed}');
      } else if (!_bytesEqual(result.content, localContent)) {
        downloadFailures.add('$cid: content mismatch with local file');
      }
    } catch (e) {
      downloadFailures.add('$cid: $e');
    }
  }

  var hasErrors = false;

  if (mismatches.isNotEmpty) {
    stdout.writeln('Found CID mismatches:');
    for (final mismatch in mismatches) {
      stdout.writeln('- $mismatch');
    }
    hasErrors = true;
  }

  if (downloadFailures.isNotEmpty) {
    stderr.writeln('Found download validation failures:');
    for (final failure in downloadFailures) {
      stderr.writeln('- $failure');
    }
    hasErrors = true;
  }

  if (hasErrors) {
    exitCode = 1;
    return;
  }

  stdout.writeln('All ${entries.length} CID files match their contents.');
  stdout.writeln('All ${entries.length} downloaded CIDs are valid.');
}

bool _bytesEqual(Uint8List a, List<int> b) {
  if (a.length != b.length) return false;
  for (var i = 0; i < a.length; i++) {
    if (a[i] != b[i]) return false;
  }
  return true;
}
